Implement Display and Debug for Builtins

0awful opened the pull request:
Closes #110.

GodotRust commented:
API docs are being generated and will be shortly available at: https://godot-rust.github.io/docs/gdext/pr-553

0awful commented:
Somehow this picked up a bunch of other people's changes. I must've messed up the `git rebase -i` `git push -f origin <branch-name>`.

0awful commented:
> Somehow this picked up a bunch of other people's changes. I must've messed up the `git rebase -i` `git push -f origin <branch-name>`.

Cleaned this up, but now have a different commit from merging in master

0awful commented:
Got it working. :+1:

Bromeon commented on godot-core/src/builtin/array.rs:
````diff
@@ -675,6 +675,26 @@ impl<T: GodotType> fmt::Debug for Array<T> {
+impl<T: GodotType + fmt::Display> fmt::Display for Array<T> {
+    /// Formats `Array` to match Godot's string representation
+    ///
+    /// Example:
+    /// ```no_run
+    /// # use godot::prelude::*;
+    /// let array = array![1,2,3,4];
+    /// assert_eq!(format!("{}", array), "[1, 2, 3, 4]");
````
This misses closing code tags. "Example" could also be a section title like elsewhere.
I also made things a bit more compact and less ambiguous with "array" identifiers:
```suggestion
    /// # Example
    /// ```no_run
    /// # use godot::prelude::*;
    /// let a = array![1, 2, 3, 4];
    /// assert_eq!(format!("{a}"), "[1, 2, 3, 4]");
    /// ```
```

Bromeon commented on godot-core/src/builtin/array.rs:
```diff
@@ -675,6 +675,26 @@ impl<T: GodotType> fmt::Debug for Array<T> {
+            if count != 0 {
+                write!(f, ", ")?;
+            }
+            write!(f, "{}", v)?;
```
```suggestion
            write!(f, "{v}")?;
```

Bromeon replied:
Btw: is this ChatGPT code? 😁

0awful replied:
Not at all :joy:. Just followed the patterns of rust by example and the std library docs.

Bromeon replied:
Ah, then ChatGPT stole it from there 😅

Bromeon commented on godot-core/src/builtin/array.rs:
```diff
@@ -675,6 +675,26 @@ impl<T: GodotType> fmt::Debug for Array<T> {
 }
 
+impl<T: GodotType + fmt::Display> fmt::Display for Array<T> {
+    /// Formats `Array` to match Godot's string representation
```
```suggestion
    /// Formats `Array` to match Godot's string representation.
```

(We should possibly change this elsewhere too, but we adopted this convention of full sentences only after some time).

0awful replied:
Updating all instances in my comments to have the full sentences paradigm.

Bromeon commented on itest/rust/src/builtin_tests/containers/array_test.rs:
```diff
@@ -452,6 +452,12 @@ fn untyped_array_try_from_typed() {
     node.free();
 }
 
+#[itest]
+fn array_should_format_with_display() {
+    assert_eq!("[1, 2, 3, 4]", format!("{}", array![1, 2, 3, 4]));
+    assert_eq!("[]", format!("{}", Array::<real>::new()));
+}
```
Could you use `assert_eq!(actual, expected)` convention like existing tests? I.e. swap arguments.

Also don't hesitate to create some "actual" variables for readability.

0awful replied:
Sounds good

Bromeon commented on itest/rust/src/builtin_tests/containers/dictionary_test.rs:
```diff
@@ -615,3 +615,19 @@ fn dictionary_iter_erase() {
      */
 }
+
+#[itest]
+fn dictionary_should_format_with_display() {
+    assert_eq!(format!("{}", Dictionary::new()), "{  }");
+    assert_eq!(
+        format!(
+            "{}",
+            dict! {
+                "one": 1,
+                "two": true,
+                "three": Variant::nil()
+            }
+        ),
+        "{ one: 1, two: true, three: <null> }"
+    )
+}
```
Also this would benefit from extracting expressions into variables. Ideally the assert statements are single-line and not nested.

0awful replied:
Will Do.

Bromeon commented on godot-core/src/builtin/dictionary.rs:
```diff
@@ -308,6 +308,19 @@ impl fmt::Debug for Dictionary {
         write!(f, "{:?}", self.to_variant().stringify())
     }
 }
+impl fmt::Display for Dictionary {
+    /// Formats `Dictionary` to match Godot's string representation
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{{ ")?;
+        for (count, entry) in self.iter_shared().enumerate() {
+            if count != 0 {
+                write!(f, ", ")?;
+            }
+            write!(f, "{}: {}", entry.0, entry.1)?;
+        }
+        write!(f, " }}")
+    }
+}
```
Are the spaces after `{ ` and before ` }` consistent with the way Godot formats dictionaries?

0awful replied:
Yes, Its quite strange but it is how an empty dictionary is logged out.

0awful replied:
Particularly strange because arrays don't do that.

Bromeon reviewed (commented):
Thanks a lot!

0awful commented:
Added `Closes #110` and made changes called out by @Bromeon 
1. Switched ordering of assert_eq!() to always use (actual, expected) semantics
2. Broke out the setup for test cases to allow for 1 line assert_eq!() statements
3. Changed write!() macro invocations to take advantage of named variables ("{a}") over ("{}", a)
4. Added periods at the end of sentences
5. Made sure the code block was closed

Bromeon commented:
Thanks a lot!
